pkg/core/bundle_list.go

```diff
@@ -12,52 +13,128 @@ import (
 )
 
 const (
-	maxBundlesToList = 1000000
+	maxBundlesToList  = 1000000
+	typicalBundlesNum = 1000
 )
 
+type bundleChan struct {
+	bundle model.BundleDescriptor
+	err    error
+}
+
+type bundlesChan struct {
+	bundles []model.BundleDescriptor
+	err     error
+}
+
+// ListBundles returns the list of bundle descriptors from a repo.
+//
 // TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store) ([]model.BundleDescriptor, error) {
+func ListBundles(repo string, store storage.Store, opts ...BundleOption) ([]model.BundleDescriptor, error) {
+	var concurrencyFactor int
+	b := defaultBundle()
+	for _, bApply := range opts {
+		bApply(&b)
+	}
+	concurrencyFactor = b.concurrentBundleList
+
 	// Get a list
 	e := RepoExists(repo, store)
 	if e != nil {
 		return nil, e
 	}
+
+	// this call is not made async yet
 	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxBundlesToList)
 	if err != nil {
 		return nil, err
 	}
-	bds := make([]model.BundleDescriptor, 0)
-	for _, k := range ks {
+
+	var wg sync.WaitGroup
+
+	result := make(chan bundlesChan)
+	bundle := make(chan bundleChan, concurrencyFactor)
+
+	wg.Add(1)
+	go func(bundle <-chan bundleChan, result chan<- bundlesChan, wg *sync.WaitGroup) {
+		// coalesce result
```

Comment:
why not coalesce the result synchronously at the end of `ListBundles`?  the `bundle` channel closes when all `getBundleAsync` goroutines finish, and i get that it's necessary for `ListBundles` to not leak goroutines...

yet what's the harm in setting a `var errRes error` (or probably some better name) in the scope of `ListBundles` and possibly setting that to non-`nil` in a synchronous `for .. range bundles` at the end of the function?  then `return bds, errRes`, why not?  i don't suppose there's any harm in returning a non-empty slice of bundle descriptors in an error case, although a conditional around the `return` could resolve even that if it's a worry. 

it is at most marginally clearer imo to coalesce results in a separate goroutine, and additionally, it [costs a fair amount of time](https://syslog.ravelin.com/so-just-how-fast-are-channels-anyway-4c156a407e45) to push bytes through a channel.  in the case of the `bundle` channel, the time to push bytes is worth it because of the parallelism of reads from gcs and deserialization.  here, my opinion (at least at the moment) is that any code style improvement this goroutine might offer is not worth the computational time to push bytes through the `result` channel.

Reply:
Extra cost of pushing through channels is a valid point I did not consider.
Initial design (2 days ago) was pushing keys and coalescing results synchronously but then I met a blocking situation. Basically,  your advice of launching the workers _before_ is probably the right solution (only justification was to avoid looping twice).